File: apps/api/src/clients/rabbitmq.ts
import amqp from 'amqplib'
import type { Channel, ChannelModel } from 'amqplib'
import { config } from '../config/index.js'

let connection: ChannelModel | null = null
let channel: Channel | null = null

const EXCHANGE = 'football_oracle'

export async function connectRabbitMQ(): Promise<Channel> {
  if (channel) return channel

  // Retry logic
  let retries = 10
  while (retries > 0) {
    try {
      connection = await amqp.connect(config.rabbitmq.url)
      channel = await connection.createChannel()
      await channel.assertExchange(EXCHANGE, 'topic', { durable: true })
      console.log('[RabbitMQ] Connected and exchange created')
      return channel
    } catch (error) {
      retries--
      console.log(`[RabbitMQ] Connection failed, retrying... (${retries} left)`)
      await new Promise((r) => setTimeout(r, 2000))
    }
  }
  throw new Error('Could not connect to RabbitMQ')
}

export async function publish(routingKey: string, message: object): Promise<void> {
  if (!channel) throw new Error('RabbitMQ not connected')

  const content = Buffer.from(JSON.stringify(message))
  channel.publish(EXCHANGE, routingKey, content, { persistent: true })

  console.log(`[RabbitMQ] Published to ${routingKey}:`, message)
}

export async function subscribe(
  routingKey: string,
  handler: (message: object) => Promise<void>
): Promise<void> {
  if (!channel) throw new Error('RabbitMQ not connected')

  const queue = await channel.assertQueue('', { exclusive: true })
  await channel.bindQueue(queue.queue, EXCHANGE, routingKey)

  console.log(`[RabbitMQ] Subscribed to ${routingKey}`)

  channel.consume(queue.queue, async (msg) => {
    if (msg) {
      const content = JSON.parse(msg.content.toString())
      console.log(`[RabbitMQ] Received from ${routingKey}:`, content)
      await handler(content)
      channel!.ack(msg)
    }
  })
}

export async function closeRabbitMQ(): Promise<void> {
  if (channel) await channel.close()
  if (connection) await connection.close()
}

File: services/scraper/src/clients/rabbitmq.py
import json
import pika
from urllib.parse import urlparse
from src.config import RABBITMQ_URL, EXCHANGE

connection = None
channel = None


def connect():
    global connection, channel

    params = pika.URLParameters(RABBITMQ_URL)
    connection = pika.BlockingConnection(params)
    channel = connection.channel()

    channel.exchange_declare(exchange=EXCHANGE, exchange_type="topic", durable=True)

    print("[RabbitMQ] Connected and exchange ready")
    return channel


def publish(routing_key: str, message: dict):
    global channel
    if not channel:
        raise Exception("RabbitMQ not connected")

    body = json.dumps(message)
    channel.basic_publish(
        exchange=EXCHANGE,
        routing_key=routing_key,
        body=body,
        properties=pika.BasicProperties(delivery_mode=2),
    )
    print(f"[RabbitMQ] Published to {routing_key}: {message}")


def subscribe(routing_key: str, handler):
    global channel
    if not channel:
        raise Exception("RabbitMQ not connected")

    result = channel.queue_declare(queue="", exclusive=True)
    queue_name = result.method.queue

    channel.queue_bind(exchange=EXCHANGE, queue=queue_name, routing_key=routing_key)

    print(f"[RabbitMQ] Subscribed to {routing_key}")

    def callback(ch, method, properties, body):
        message = json.loads(body)
        print(f"[RabbitMQ] Received from {routing_key}: {message}")
        handler(message)
        ch.basic_ack(delivery_tag=method.delivery_tag)

    channel.basic_consume(queue=queue_name, on_message_callback=callback)


def start_consuming():
    global channel
    print("[RabbitMQ] Waiting for messages...")
    channel.start_consuming()


def close():
    global connection
    if connection:
        connection.close()
